### ISSUE TITLE
Incorrect syntax highlighting for C++ constructor split across lines

### ISSUE THREAD

**Issue by sunnyps** [sublimehq/Packages#360]
Ubuntu 14.04 with ST3 3112.

The syntax highlighting (Monokai) breaks when a C++ constructor is split across multiple lines e.g. with "A::B::C::\nD() {}" the first line "A::B::C::" is not highlighted correctly but the second line "D() {}" is.

**Comment by kleinfreund** [comment]
The `D` is not being recognized as a constructor at all as far as I can tell. If you press <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>Alt</kbd>+<kbd>P</kbd> while having `D` selected will not show `source.c++ meta.function.c++ entity.name.function.constructor.c++` but `source.c++ meta.function.parameters.c++ meta.group.c++ punctuation.definition.group.begin.c++`.

```c++
A::B::C::D() {}
//    ^ entity.name.function.constructor.c++
//       ^ punctuation.definition.group.begin.c++
```

**Comment by wbond** [comment]
This feels slightly contrived, and is going to be hard to get 100% right. Seeing part of namespace path ending in `::`before a newline could be a return type name or could be a constructor method name.

Do you have an example of a real, openly-viewable codebase that uses this pattern? It would be helpful in trying to see if there are heuristics that we could use to try and get the most code correctly highlighted. Unfortunately we won't be able to correct highlight C/C++ 100% of the time due to the preprocessor, but we are working on trying to get reasonable codebases.

Lexing is done on a line-by-line basis for performance reasons, so we can't look ahead for N lines to see if a certain pattern appears.

**Comment by sunnyps** [comment]
Here's an example: https://code.google.com/p/chromium/codesearch#chromium/src/cc/output/shader.cc&l=1750

To clarify the components A, B, C and D were not namespaces but nested classes. Also the correct example should be something like A::B::C::\nC () {} where C is both the class name and the constructor.

**Comment by wbond** [comment]
This case should be resolved by bc618c6762a150bc0576f6d1a6bb3a6928b466f1.

I did not implement any other instances of path wrapping since it gets complicated quickly, and we need to implement it based on heuristics.

**Comment by sunnyps** [comment]
Thanks!
